### PR TITLE
Fixes UB when using legacy python functions and mark_non_differentiable

### PR DESCRIPTION
**Pull request by zou3519**
This should fix the flaky segfaults I was seeing in `test_autograd/test_dep_nograd`.

If an output of a python Function is marked as non_differentiable,
autograd won't save a gradfn for that output. During the backward
pass, this translates to autograd passing an undefined tensor to the
backward of the Function. The legacy python Function path checks
if *any* of the inputs to backward requires_grad, here:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/functions/utils.cpp#L16

This requires_grad check uses Variable::get(), which casts the
undefined tensor to a VariableImpl and then accesses the _requires_grad
member. This is UB because the undefined tensor is NOT a VariableImpl.

The fix I did here is to add a check for if the variable/tensor is defined
in the legacy python Function code path.

The other that that I could do is to reevaluate how autograd treats outputs marked as non_differentiable. The inputs to the legacy python function are already zero-filled, but they don't get zero-filled early enough to avoid `requires_grad` being called on them right now.

cc @ezyang 

### Test Plan
I should probably figure out how to use ASAN and run my code on an ASAN build... TBD in progress.

### CODE REVIEW

**Comment by apaszke** [comment]
Wouldn't it be better to move [this block](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_function.cpp#L107-L110) below [that part](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_function.cpp#L112-L126)? Should be enough to ensure all inputs are defined

**Comment by zou3519** [comment]
@apaszke The problem is that `legacy_apply` calls [wrap_outputs](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_function.cpp#L90), which looks at the C++ inputs (not the Python inputs) and checks the C++ inputs for requires_grad. 

The Python inputs are okay because there's some code later that zeros them out before calling the Python function on them (and we could simply that logic by doing the change you suggested, moving that code block)

File: torch/csrc/autograd/function.h [hunk 1]
```diff
@@ -56,7 +56,7 @@ struct MakeNextFunctionList : IterArgs<MakeNextFunctionList> {
 inline bool any_variable_requires_grad(const variable_list& variables) {
   return std::any_of(
       variables.begin(), variables.end(), [](const Variable& variable) {
-        return variable.requires_grad();
+        return variable.defined() and variable.requires_grad();
```
Comment (goldsborough) [inline torch/csrc/autograd/function.h]: Did you mean to use `and` instead of `&&`?

Comment (zou3519) [reply torch/csrc/autograd/function.h]: `and` is equivalent to `&&` I believe

**Comment by apaszke** [comment]
@zou3519 I understand, but isn’t it caused by undefined inputs? If you were to swap those two blocks you would have a guarantee that they will all be defined. I think it’s a better fix that will let us avoid unnecessary checks in other places

File: torch/csrc/autograd/function.h (hunk 1 above)
Comment (goldsborough) [reply torch/csrc/autograd/function.h]: Yeah but we don't use it anywhere else, and [no one uses it in practice](https://stackoverflow.com/questions/14740864/any-downside-to-using-the-and-operator-vs-the-operator)

Comment (apaszke) [reply torch/csrc/autograd/function.h]: We use it pretty much everywhere ;)

**Comment by zou3519** [comment]
@apaszke If I'm understanding the block here correctly https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/python_function.cpp#L112-L126 , I think it ensures that the python inputs are defined, but doesn't change the C++ inputs. If I swapped the two blocks, we'd guarantee the python inputs (pyInputs) has defined tensors, but the C++ inputs (inputs) are still undefined after the block (and those are what get passed to wrap_outputs).

I could change `PyFunction::legacy_apply` to not call `wrap_outputs` to process the outputs and instead follow how`PyFunction::apply` process its outputs to fix this to avoid the checks, though. I'll look into that more

**Comment by ezyang** [comment]
I confirmed that this solved the ASAN problem locally for me. CI for ASAN is being tracked at #5283

**Pull request merged**